python_ta/checkers/possibly_undefined_checker.py

```diff
@@ -0,0 +1,157 @@
+            if isinstance(statement, astroid.FunctionDef):
+                continue
+            for node in self._get_child_nodes(statement, (astroid.AssignName, astroid.DelName),
+                                              astroid.FunctionDef):
+                if isinstance(node, astroid.AssignName):
+                    gen.add(node.name)
+                else:
+                    kill.add(node.name)
+            self._analyze_statement(statement, gen.union(in_facts.difference(kill)), all_assigns)
+        return gen.union(in_facts.difference(kill))
+
+    def _analyze_statement(self, statement: NodeNG, facts: Set[str], local_vars: Set[str]) -> None:
+        nodes = statement.nodes_of_class(astroid.Name)
+        for node in nodes:
+            name = node.name
+            if not self._is_function_name(node) and name in local_vars and name not in facts:
+                self._possibly_undefined.add(node)
+            elif node in self._possibly_undefined:
+                self._possibly_undefined.remove(node)
+
+    def _get_assigns(self, node: Union[astroid.FunctionDef, astroid.Module]) -> Set[str]:
+        """Returns a set of all local and parameter variables that could be
+        defined in the program (either a function or module).
+
+        IF a variable 'v' is defined in a function and there is no global/nonlocal
+        statement applied to 'v' THEN 'v' is a local variable.
+
+        Note that `local variable` in the context of a module level analysis,
+        refers to global variables.
+        """
+        assigns = set()
+        kills = set()
+        for statement in self._get_child_nodes(node, (astroid.AssignName, astroid.Global,
+                                                      astroid.node_classes.Nonlocal),
+                                               astroid.FunctionDef):
+            if isinstance(statement, astroid.AssignName):
+                assigns.add(statement.name)
+            elif isinstance(statement, (astroid.Global, astroid.Nonlocal)):
+                for name in statement.names:
+                    kills.add(name)
+
+        return assigns.difference(kills)
+
+    def _is_function_name(self, node: astroid.Name) -> bool:
+        return isinstance(node.parent, astroid.Call) and node == node.parent.func
+
+    def _get_blocks_po(self, node: Union[astroid.Module, astroid.FunctionDef]) -> List[CFGBlock]:
+        """Return the sequence of all blocks in this graph in the order of
+        a post-order traversal."""
+        return self._get_blocks(node.cfg_block, set())
+
+    def _get_blocks(self, block: CFGBlock, visited) -> List[CFGBlock]:
+        if block.id in visited:
+            return []
+
+        visited.add(block.id)
+        blocks = []
+
+        for succ in block.successors:
+            blocks.extend(self._get_blocks(succ.target, visited))
+        blocks.append(block)
+
+        return blocks
+
+    def _get_child_nodes(self, node: NodeNG, klasses, skip_klasses=None):
```

Comment:
This method has the absolute same implementation as the `NodeNG.nodes_of_class` method. The reason I chose to rewrite the method here is because its public interface didn't support multiple classes to select (or ignore) even though the implementation apparently does. I think it shouldn't be a problem to actually use the `nodes_of_class` method since it's highly improbable that the creators of astroid will change the implementation, however, I feel that with an explicit implementation we are more certain that our code will work the way we intend it to. What do you think? @david-yz-liu

Reply:
Great point! I think it's fine to use the `nodes_of_class` method. Pylint itself has a couple of instances of using it with multiple classes. If you want to, you could submit an issue or pull request in the astroid repo to update the documentation. (Note that `skip_klasses` can also be a tuple.)